x/liquidstaking/keeper/derivative.go

```diff
@@ -0,0 +1,163 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	vesting "github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/kava-labs/kava/x/liquidstaking/types"
+)
+
+// MintDerivative mints a new derivative
+func (k Keeper) MintDerivative(ctx sdk.Context, delegatorAddr sdk.AccAddress, valAddr sdk.ValAddress, coin sdk.Coin) error {
+
+	_, found := k.stakingKeeper.GetValidator(ctx, valAddr)
+	if !found {
+		return types.ErrNoValidatorFound
+	}
+
+	validator, found := k.stakingKeeper.GetValidator(ctx, valAddr)
+	if !found {
+		return types.ErrNoValidatorFound
+	}
+
+	delegation, found := k.stakingKeeper.GetDelegation(ctx, delegatorAddr, valAddr)
+	if !found {
+		return types.ErrNoDelegatorForAddress
+	}
+
+	if coin.Denom != k.stakingKeeper.BondDenom(ctx) {
+		return types.ErrOnlyBondDenomAllowdForTokenize
+	}
+
+	delegationAmount := validator.Tokens.ToDec().Mul(delegation.GetShares()).Quo(validator.DelegatorShares)
+	if coin.Amount.GT(sdk.Int(delegationAmount)) {
+		return types.ErrNotEnoughDelegationShares
+	}
+
+	acc := k.accountKeeper.GetAccount(ctx, delegatorAddr)
+	if acc != nil {
+		acc, ok := acc.(vesting.VestingAccount)
+		if ok {
+			// if account is a vesting account, it checks if free delegation (non-vesting delegation) is not exceeding
+			// the tokenize share amount and execute further tokenize share process
+			// tokenize share is reducing unlocked tokens delegation from the vesting account and further process
+			// is not causing issues
+			delFree := acc.GetDelegatedFree().AmountOf(coin.Denom)
+			if delFree.LT(coin.Amount) {
+				return types.ErrExceedingFreeVestingDelegations
+			}
+		}
+	}
+
+	liquidCoinDenom := k.GetLiquidStakingTokenDenom(ctx, valAddr)
+	liquidCoin := sdk.NewCoin(liquidCoinDenom, coin.Amount)
+
+	delegationHolder, found := k.GetDelegationHolder(ctx, valAddr)
+	if !found {
+		// Generate specific module account address for this valAddr
+		modAccName := authtypes.NewModuleAddress(fmt.Sprintf("%s:%s", types.ModuleName, valAddr.String()))
+		modBaseAcc := authtypes.NewBaseAccount(modAccName, nil, 0, 0)
+		modAcc := authtypes.NewModuleAccount(modBaseAcc, types.ModuleName, []string{}...)
+		k.accountKeeper.SetModuleAccount(ctx, modAcc)
+
+		delegationHolder = types.NewDelegationHolder(modAcc.GetAddress(), valAddr, liquidCoin)
+	} else {
+		// Check that the existing delegation liquid staking token denom matches
+		if delegationHolder.Delegations.Denom != liquidCoinDenom {
+			return types.ErrInvalidLiquidCoinDenom
+		}
+		// Add the liquid staking coin to the delegations
+		delegationHolder.Delegations = delegationHolder.Delegations.Add(liquidCoin)
```

Comment:
I think we don't need to track the total in this module as the delegation object will contain the total shares converted to derivatives, which should be the same as the number of derivative tokens.

Reply:
https://github.com/Kava-Labs/kava/pull/1271/commits/200e9785c49f4522180545dba6744b42d3691f72